extension/datadogextension/internal/httpserver/httpserver.go

```diff
@@ -0,0 +1,178 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package httpserver // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/datadogextension/internal/httpserver"
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/comp/forwarder/defaultforwarder"
+	"github.com/DataDog/datadog-agent/pkg/serializer/marshaler"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/datadogextension/internal/payload"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/datadog/agentcomponents"
+)
+
+var nowFunc = time.Now
+
+// Server provides local metadata server functionality (display the otel_collector payload locally) as well as
+// functions to serialize and export this metadata to Datadog backend.
+type Server struct {
+	// server is used to respond to local metadata requests
+	server *http.Server
+	// logger is passed from the extension to allow logging
+	logger *zap.Logger
+	// serializer is a datadog-agent component used to forward payloads to Datadog backend
+	serializer agentcomponents.SerializerWithForwarder
+	// config contains the httpserver configuration values
+	config *Config
+
+	// payload is the metadata to send to Datadog backend
+	// TODO: support generic payloads
+	payload payload.OtelCollectorPayload
+
+	// mu protects concurrent access to the serializer
+	// Note: Only protects serializer operations, not field reads since they're set once during initialization
+	mu sync.Mutex
+}
+
+// NewServer creates a new HTTP server instance.
+// It should be called after NotifyConfig has received full configuration.
+// TODO: support generic payloads
+func NewServer(
+	logger *zap.Logger,
+	s agentcomponents.SerializerWithForwarder,
+	config *Config,
+	hostname string,
+	uuid string,
+	p payload.OtelCollector,
+) *Server {
+	oc := payload.OtelCollectorPayload{
+		Hostname:  hostname,
+		Timestamp: nowFunc().UnixNano(),
```

Comment:
Before, IIRC, we were updating the timestamp on each call to `SendPayload`, should we still do that?

Reply:
ah, missed that, sorry!